spring-rabbit-stream/src/main/java/org/springframework/rabbit/stream/support/converter/StreamMessageConverter.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.rabbit.stream.support.converter;
+
+import java.util.Map;
+
+import org.springframework.amqp.core.Message;
+import org.springframework.amqp.core.MessageBuilder;
+import org.springframework.amqp.core.MessageProperties;
+import org.springframework.amqp.support.converter.MessageConversionException;
+import org.springframework.amqp.support.converter.MessageConverter;
+import org.springframework.amqp.utils.JavaUtils;
+import org.springframework.util.Assert;
+
+import com.rabbitmq.stream.MessageBuilder.PropertiesBuilder;
+import com.rabbitmq.stream.Properties;
+import com.rabbitmq.stream.codec.WrapperMessageBuilder;
+
+/**
+ * @author Gary Russell
+ * @since 2.4
+ *
+ */
+public class StreamMessageConverter implements MessageConverter {
+
+	@Override
+	public Message toMessage(Object object, MessageProperties messageProperties) throws MessageConversionException {
+		Assert.isInstanceOf(com.rabbitmq.stream.Message.class, object);
+		com.rabbitmq.stream.Message qpid = (com.rabbitmq.stream.Message) object;
+		toMessageProperties(qpid, messageProperties);
+		return MessageBuilder.withBody(qpid.getBodyAsBinary()).andProperties(messageProperties).build();
+	}
+
+	@Override
+	public com.rabbitmq.stream.Message fromMessage(Message message) throws MessageConversionException {
+		WrapperMessageBuilder builder = new WrapperMessageBuilder();
```

Comment:
Where will the created message typically go? If it's meant to be published, it would be more optimized to use a `MessageBuilder` from the environment `Codec`. The codec's `MessageBuilder` will directly use an AMQP 1.0 message (QPid by default) under the covers. If passed in to a publisher, the resulting message will be converted to bytes and go on the wire.

The `WrapperMessageBuilder` will create a wrapper, that will store values (body, properties), and then will get converted to AMQP 1.0 message format before being sent on the wire if published. This is an extra step compared to the previous case.

Reply:
I haven't handled the producer side so this code is not currently used, but I will take your advce.

Reply:
@acogoluegnes How to get the codec from the environment? It's not on the interface and is package-protected on the implementation.

Thanks.

Reply:
Oh, I see from the doc `environment -> producer -> messageBuilder` (which comes from the environment's codec).

Typically, our message converters are stateless; I'd rather not tie it to a particular producer; it would be helpful if we could get the message builder directly from the environment.

Reply:
I'd suggest to declare a codec singleton that would be injected in the environment and in the message converter. The message converter could do without the codec, by using an internal `Supplier<MessageBuilder>` that creates a new instance of  `WrapperMessageBuilder`. When created with a codec the message converter would use the codec in the `Supplier<MessageBuilder>`.